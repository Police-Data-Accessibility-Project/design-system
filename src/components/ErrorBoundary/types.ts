import { ComponentPublicInstance } from 'vue';

export interface PdapErrorEmitted {
	error: Error;
	vm: ComponentPublicInstance | null;
	info?: string;
}

export interface PdapErrorBoundaryProps {
	component: string;
	onError?: (
		error: Error,
		target?: ComponentPublicInstance | null | undefined,
		info?: string
	) => void;
	params?: Record<string, string>;
}
