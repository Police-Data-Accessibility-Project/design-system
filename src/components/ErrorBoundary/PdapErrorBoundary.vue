<template>
	<slot v-if="!error$" />
	<component
		:is="component"
		v-else
		class="pdap-flex-container-center h-[full]"
		:v-bind="params"
	>
		<h1>Oops, something went wrong!</h1>
		<p class="max-w-full" data-test="error-boundary-message">
			If you keep seeing this message, please email
			<a href="mailto:contact@pdap.io">contact@pdap.io</a> for assistance.
		</p>
	</component>
</template>

<script setup lang="ts">
import { ComponentPublicInstance, onErrorCaptured, ref } from 'vue';
import { PdapErrorBoundaryProps } from './types';

const props = withDefaults(defineProps<PdapErrorBoundaryProps>(), {
	component: 'div',
	onError: undefined,
	params: undefined,
	stopPropagation: false,
});

interface PdapErrorEmitted {
	error: Error;
	vm: ComponentPublicInstance | null;
	info?: string;
}

const emits = defineEmits<{
	(event: 'error', { error, vm, info }: PdapErrorEmitted): void;
}>();
const error$ = ref<Error>();
const info$ = ref<string | undefined>('');

function interceptError(
	error: Error,
	vm: ComponentPublicInstance | null,
	info?: string
) {
	error$.value = error;
	info$.value = info;

	props.onError?.(error, vm);
	emits('error', { error: error, vm, info });

	if (props.stopPropagation) return false;
}

/* Impossible and unwise to try testing this, so we remove from the coverage report */
/* c8 ignore next 3 */
onErrorCaptured((err, vm, info) => {
	interceptError(err, vm, info);
});
</script>
