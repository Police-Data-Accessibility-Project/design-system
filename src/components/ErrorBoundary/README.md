# ErrorBoundary
Intercepts uncaught errors from its children and renders an error UI in place of its children.

## Props
| name        | required? | types    | description                     | default |
| ----------- | --------- | -------- | ------------------------------- | ------- |
| `component` | no        | `string` | component to render as fallback | `'div'` |

## Example
_From data-sources `App.vue`: This will catch any uncaught error at the route level and render the error fallback_

```
<template>
	<AuthWrapper>
		<Header :logo-image-src="lockup" />
		<ErrorBoundary component="main">
			<router-view />
		</ErrorBoundary>
		<Footer :logo-image-src="acronym" />
	</AuthWrapper>
</template>
```

_From data-sources `SearchResultPage.vue`: For each of these results, the error boundary will catch any uncaught error only within its `SearchResultCard` component child and render the error fallback. In this case, the error will_ not _bubble any further up the DOM tree, allowing us to render the rest of the UI that did not error._

```
<div class="grid pdap-grid-container-column-3 gap-4">
  <ErrorBoundary
    v-for="result in [...getAllRecordsFromSection(section)]"
    :key="result.type"
  >
    <SearchResultCard
      data-test="search-results-cards"
      :data-source="result"
    />
  </ErrorBoundary>
</div>
```