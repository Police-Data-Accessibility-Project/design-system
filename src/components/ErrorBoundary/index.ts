export { default as ErrorBoundary } from './PdapErrorBoundary.vue';
