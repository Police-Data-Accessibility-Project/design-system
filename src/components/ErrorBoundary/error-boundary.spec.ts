import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, it } from 'vitest';
import ErrorBoundary from './PdapErrorBoundary.vue';
import { nextTick } from 'vue';

let wrapper;

describe('ErrorBoundary', () => {
	beforeEach(() => {
		wrapper = mount(ErrorBoundary, {
			slots: {
				default: '<div data-test="default-slot" />',
			},
		});
	});

	it('renders slot content with no error', () => {
		expect(wrapper.find('[data-test="default-slot"]').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('renders error content with error', async () => {
		wrapper.vm.interceptError(new Error('Generic error'));
		await nextTick();

		expect(wrapper.find('[data-test="error-boundary-message"]').exists()).toBe(
			true
		);
		expect(wrapper.html()).toMatchSnapshot();
	});
});
