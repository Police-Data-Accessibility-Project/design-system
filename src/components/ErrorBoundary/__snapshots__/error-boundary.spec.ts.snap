// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary > renders error content with error 1`] = `
<div class="pdap-flex-container-center h-[full]">
  <h1>Oops, something went wrong!</h1>
  <p class="max-w-full"> If you keep seeing this message, please email <a href="mailto:contact@pdap.io">contact@pdap.io</a> for assistance. </p>
</div>
`;

exports[`ErrorBoundary > renders slot content with no error 1`] = `<div></div>`;
